File: api/numerology.py
"""
Pythagorean Numerology Calculator.
Pure math — 100% accurate.
"""

from datetime import datetime


def _digit_sum(n):
    while n > 9 and n not in (11, 22, 33):
        n = sum(int(d) for d in str(n))
    return n


LETTER_VALUES = {c: i for i, chars in enumerate(
    ["abc", "def", "ghi", "jkl", "mno", "pqr", "stu", "vwx", "yz"], 1)
    for c in chars}


def _name_sum(name, filter_fn=None):
    cleaned = "".join(c.lower() for c in name if c.isalpha())
    if filter_fn:
        cleaned = "".join(c for c in cleaned if filter_fn(c))
    if not cleaned:
        return 0
    return _digit_sum(sum(LETTER_VALUES.get(c, 0) for c in cleaned))


PINNACLE_MAP = {
    1: "Independence and self-discovery. Forge your own path.",
    2: "Cooperation and relationships. Develop patience.",
    3: "Creative self-expression. Share your gifts.",
    4: "Build foundations. Hard work pays off.",
    5: "Freedom and change. Stay adaptable.",
    6: "Responsibility and service. Nurture others.",
    7: "Inner growth and spiritual development.",
    8: "Material achievement and recognition.",
    9: "Completion and humanitarian service. Give back.",
    11: "Master illumination cycle. Heightened intuition.",
    22: "Master builder cycle. Large-scale manifestation.",
    33: "Master teacher cycle. Unconditional love.",
}

CHALLENGE_MAP = {
    0: "Varied life tests — no single dominant challenge.",
    1: "Develop independence, overcome self-doubt.",
    2: "Overcome oversensitivity, build confidence.",
    3: "Avoid scattering energy, develop focus.",
    4: "Release rigidity, learn flexibility.",
    5: "Avoid escapism, channel freedom constructively.",
    6: "Balance helping others with self-care.",
    7: "Overcome isolation, stay connected.",
    8: "Integrate power with compassion.",
}


def calculate_numerology(birth_date, name=""):
    """Calculate Pythagorean numerology numbers.

    Args:
        birth_date: "YYYY-MM-DD"
        name: Full birth name (optional)
    """
    y, m, d = map(int, birth_date.split("-"))

    # Life Path
    life_path = _digit_sum(_digit_sum(m) + _digit_sum(d) + _digit_sum(y))

    # Birthday
    birthday = _digit_sum(d)

    # Name-based numbers
    expression = _name_sum(name) if name else 0
    soul_urge = _name_sum(name, lambda c: c in "aeiou") if name else 0
    personality = _name_sum(name, lambda c: c not in "aeiou") if name else 0

    # Pinnacle cycles
    first = _digit_sum(m + d)
    second = _digit_sum(d + y)
    third = _digit_sum(first + second)
    fourth = _digit_sum(m + y)

    base_age = 36 - life_path
    pinnacles = [
        {"age": f"0-{base_age}", "number": first, "meaning": PINNACLE_MAP.get(first, "")},
        {"age": f"{base_age}-{base_age + 9}", "number": second, "meaning": PINNACLE_MAP.get(second, "")},
        {"age": f"{base_age + 9}-{base_age + 18}", "number": third, "meaning": PINNACLE_MAP.get(third, "")},
        {"age": f"{base_age + 18}+", "number": fourth, "meaning": PINNACLE_MAP.get(fourth, "")},
    ]

    # Challenge cycles
    challenges = [
        {"age": f"0-{base_age}", "number": _digit_sum(abs(m - d))},
        {"age": f"{base_age}-{base_age + 9}", "number": _digit_sum(abs(d - y))},
        {"age": f"{base_age + 9}-{base_age + 18}", "number": _digit_sum(abs(first - second))},
        {"age": f"{base_age + 18}+", "number": _digit_sum(abs(m - y))},
    ]

    # Personal year/month/day
    now = datetime.now()
    py = _digit_sum(_digit_sum(d) + _digit_sum(m) + _digit_sum(now.year))
    pm = _digit_sum(py + (now.month))
    pd = _digit_sum(pm + now.day)

    return {
        "life_path": life_path,
        "expression": expression,
        "soul_urge": soul_urge,
        "personality": personality,
        "birthday": birthday,
        "pinnacle_cycles": pinnacles,
        "challenge_cycles": challenges,
        "personal_year": py,
        "personal_month": pm,
        "personal_day": pd,
    }

File: src/lib/astro/numerology.ts
// ─── Numerology Calculator ───────────────────────────────────────────
// Pythagorean numerology from birth date and name.
// 100% accurate — pure date math.

export interface NumerologyResult {
  life_path: number;
  expression: number;
  soul_urge: number;
  personality: number;
  birthday: number;
  pinnacle_cycles: { age: string; number: number; meaning: string }[];
  challenge_cycles: { age: string; number: number }[];
  personal_year: number;
  personal_month: number;
  personal_day: number;
}

function digitSum(n: number): number {
  while (n > 9 && n !== 11 && n !== 22 && n !== 33) {
    n = String(n).split("").reduce((s, d) => s + parseInt(d), 0);
  }
  return n;
}

const LETTER_VALUES: Record<string, number> = {
  a:1,b:2,c:3,d:4,e:5,f:6,g:7,h:8,i:9,
  j:1,k:2,l:3,m:4,n:5,o:6,p:7,q:8,r:9,
  s:1,t:2,u:3,v:4,w:5,x:6,y:7,z:8,
};

function nameNumber(name: string): number {
  const cleaned = name.toLowerCase().replace(/[^a-z]/g, "");
  if (!cleaned) return 0;
  const sum = cleaned.split("").reduce((s, c) => s + (LETTER_VALUES[c] || 0), 0);
  return digitSum(sum);
}

function vowelsNumber(name: string): number {
  const vowels = name.toLowerCase().replace(/[^aeiou]/g, "");
  if (!vowels) return 0;
  return digitSum(vowels.split("").reduce((s, c) => s + (LETTER_VALUES[c] || 0), 0));
}

function consonantsNumber(name: string): number {
  const cons = name.toLowerCase().replace(/[aeiou]/g, "").replace(/[^a-z]/g, "");
  if (!cons) return 0;
  return digitSum(cons.split("").reduce((s, c) => s + (LETTER_VALUES[c] || 0), 0));
}

const PINNACLE_MEANINGS: Record<number, string> = {
  1: "A period of independence and self-discovery. Take initiative.",
  2: "Cooperation and relationships take center stage. Develop patience.",
  3: "Creative self-expression and social expansion.",
  4: "Build foundations — career, home, long-term structures.",
  5: "Freedom, change, adventure. Expect unexpected opportunities.",
  6: "Responsibility, family, service. Nurture others.",
  7: "Inner growth, study, spiritual development.",
  8: "Material achievement, power, recognition.",
  9: "Completion and humanitarian service. Let go, give back.",
  11: "Master cycle of illumination. Heightened intuition.",
  22: "Master cycle of large-scale building.",
  33: "Master cycle of teaching and unconditional love.",
};

export function calculateNumerology(birthDate: string, name: string = ""): NumerologyResult {
  const d = new Date(birthDate + "T12:00:00"); // noon to avoid timezone issues
  const month = d.getMonth() + 1;
  const day = d.getDate();
  const year = d.getFullYear();

  // Life Path
  const lifePath = digitSum(digitSum(month) + digitSum(day) + digitSum(year));

  // Birthday number
  const birthday = digitSum(day);

  // Expression / Soul Urge / Personality
  const expression = name ? nameNumber(name) : 0;
  const soulUrge = name ? vowelsNumber(name) : 0;
  const personality = name ? consonantsNumber(name) : 0;

  // Pinnacle cycles
  const firstPinnacle = digitSum(month + day);
  const secondPinnacle = digitSum(day + year);
  const thirdPinnacle = digitSum(firstPinnacle + secondPinnacle);
  const fourthPinnacle = digitSum(month + year);

  const baseAge = 36 - lifePath;
  const pinnacleCycles = [
    { age: `0-${baseAge}`, number: firstPinnacle, meaning: PINNACLE_MEANINGS[firstPinnacle] || "" },
    { age: `${baseAge}-${baseAge + 9}`, number: secondPinnacle, meaning: PINNACLE_MEANINGS[secondPinnacle] || "" },
    { age: `${baseAge + 9}-${baseAge + 18}`, number: thirdPinnacle, meaning: PINNACLE_MEANINGS[thirdPinnacle] || "" },
    { age: `${baseAge + 18}+`, number: fourthPinnacle, meaning: PINNACLE_MEANINGS[fourthPinnacle] || "" },
  ];

  // Challenge cycles
  const challengeCycles = [
    { age: `0-${baseAge}`, number: digitSum(Math.abs(month - day)) },
    { age: `${baseAge}-${baseAge + 9}`, number: digitSum(Math.abs(day - year)) },
    { age: `${baseAge + 9}-${baseAge + 18}`, number: digitSum(Math.abs(firstPinnacle - secondPinnacle)) },
    { age: `${baseAge + 18}+`, number: digitSum(Math.abs(month - year)) },
  ];

  // Personal year/month/day
  const now = new Date();
  const currentYear = now.getFullYear();
  const personalYear = digitSum(digitSum(day) + digitSum(month) + digitSum(currentYear));
  const personalMonth = digitSum(personalYear + (now.getMonth() + 1));
  const personalDay = digitSum(personalMonth + now.getDate());

  return {
    life_path: lifePath,
    expression,
    soul_urge: soulUrge,
    personality,
    birthday,
    pinnacle_cycles: pinnacleCycles,
    challenge_cycles: challengeCycles,
    personal_year: personalYear,
    personal_month: personalMonth,
    personal_day: personalDay,
  };
}
